<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Consensus NMF with BiocNMF • BiocNMF</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Consensus NMF with BiocNMF">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BiocNMF</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/BiocNMF.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/getting-started.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/consensus-nmf.html">Consensus NMF</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MicTott/BiocNMF" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Consensus NMF with BiocNMF</h1>
                        <h4 data-toc-skip class="author">Michael
Totty</h4>
            
            <h4 data-toc-skip class="date">2025-09-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/MicTott/BiocNMF/blob/HEAD/vignettes/consensus-nmf.Rmd" class="external-link"><code>vignettes/consensus-nmf.Rmd</code></a></small>
      <div class="d-none name"><code>consensus-nmf.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction-to-consensus-nmf">Introduction to Consensus NMF<a class="anchor" aria-label="anchor" href="#introduction-to-consensus-nmf"></a>
</h2>
<p>Consensus Non-negative Matrix Factorization (cNMF) addresses a
fundamental limitation of standard NMF: <strong>instability across
random initializations</strong>. While standard NMF can produce
different results with different random seeds, cNMF provides robust,
reproducible gene expression programs through consensus clustering.</p>
<div class="section level3">
<h3 id="the-stability-problem">The Stability Problem<a class="anchor" aria-label="anchor" href="#the-stability-problem"></a>
</h3>
<p>Standard NMF optimization is non-convex, leading to multiple local
optima. This means: - Different random initializations produce different
solutions - Gene expression programs may not be reproducible - Optimal k
selection becomes challenging - Biological interpretation can be
inconsistent</p>
</div>
<div class="section level3">
<h3 id="the-consensus-solution">The Consensus Solution<a class="anchor" aria-label="anchor" href="#the-consensus-solution"></a>
</h3>
<p>cNMF, developed by Kotliar et al. (eLife 2019), solves this
through:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Multiple NMF runs</strong> with different
initializations</li>
<li>
<strong>Consensus clustering</strong> to group similar gene
expression programs</li>
<li>
<strong>Stability metrics</strong> for objective k selection</li>
<li>
<strong>Robust final programs</strong> through clustering and
refitting</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MicTott/BiocNMF" class="external-link">BiocNMF</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="the-cnmf-algorithm">The cNMF Algorithm<a class="anchor" aria-label="anchor" href="#the-cnmf-algorithm"></a>
</h2>
<div class="section level3">
<h3 id="step-by-step-methodology">Step-by-Step Methodology<a class="anchor" aria-label="anchor" href="#step-by-step-methodology"></a>
</h3>
<div class="section level4">
<h4 id="multiple-nmf-factorizations">1. Multiple NMF Factorizations<a class="anchor" aria-label="anchor" href="#multiple-nmf-factorizations"></a>
</h4>
<p>For each k value, run NMF many times (typically 100-200) with
different random initializations:</p>
<pre><code>For k in k_range:
    For run in 1:n_runs:
        W[k,run], H[k,run] = NMF(X, k, random_seed=run)</code></pre>
</div>
<div class="section level4">
<h4 id="gene-expression-program-collection">2. Gene Expression Program Collection<a class="anchor" aria-label="anchor" href="#gene-expression-program-collection"></a>
</h4>
<p>Collect all gene expression programs (columns of W matrices) across
runs.</p>
</div>
<div class="section level4">
<h4 id="density-filtering">3. Density Filtering<a class="anchor" aria-label="anchor" href="#density-filtering"></a>
</h4>
<p>Remove outlier programs using local density estimation: - Calculate
pairwise cosine similarities between programs - Compute local density
for each program - Filter programs below density threshold - Reduces
noise and computational burden</p>
</div>
<div class="section level4">
<h4 id="consensus-clustering">4. Consensus Clustering<a class="anchor" aria-label="anchor" href="#consensus-clustering"></a>
</h4>
<p>Cluster similar programs using k-means: - Apply k-means clustering to
filtered programs - Group similar gene expression programs together -
Each cluster represents a consensus program</p>
</div>
<div class="section level4">
<h4 id="consensus-matrix-construction">5. Consensus Matrix Construction<a class="anchor" aria-label="anchor" href="#consensus-matrix-construction"></a>
</h4>
<p>Build consensus matrix showing co-clustering frequencies: - For each
pair of programs: how often they cluster together - Provides stability
assessment for each k value - Used for final stability metrics</p>
</div>
<div class="section level4">
<h4 id="final-program-selection">6. Final Program Selection<a class="anchor" aria-label="anchor" href="#final-program-selection"></a>
</h4>
<ul>
<li>Choose optimal k based on stability metrics</li>
<li>Extract representative programs from each cluster</li>
<li>Refit usage matrix using non-negative least squares</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="implementation-details">Implementation Details<a class="anchor" aria-label="anchor" href="#implementation-details"></a>
</h3>
<p>BiocNMF implements the exact algorithm from Kotliar et al. with these
key components:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Conceptual overview (actual implementation is in C++/optimized R)</span></span>
<span><span class="va">runConsensusNMF</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">k_range</span>, <span class="va">n_runs</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="va">k_range</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># 1. Multiple NMF runs</span></span>
<span>    <span class="va">nmf_runs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">replicate</a></span><span class="op">(</span><span class="va">n_runs</span>, <span class="fu">RcppML</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RcppML/man/nmf.html" class="external-link">nmf</a></span><span class="op">(</span><span class="va">data</span>, k <span class="op">=</span> <span class="va">k</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># 2. Collect all programs</span></span>
<span>    <span class="va">all_programs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">nmf_runs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">r</span><span class="op">)</span> <span class="va">r</span><span class="op">$</span><span class="va">w</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># 3. Density filtering</span></span>
<span>    <span class="va">filtered_programs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/applyDensityFiltering.html">applyDensityFiltering</a></span><span class="op">(</span><span class="va">all_programs</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># 4. Consensus clustering</span></span>
<span>    <span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">filtered_programs</span><span class="op">)</span>, centers <span class="op">=</span> <span class="va">k</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># 5. Build consensus matrix</span></span>
<span>    <span class="va">consensus_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildConsensusMatrix.html">buildConsensusMatrix</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># 6. Calculate stability metrics</span></span>
<span>    <span class="va">stability</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateStabilityMetrics.html">calculateStabilityMetrics</a></span><span class="op">(</span><span class="va">consensus_mat</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co"># Select optimal k and refit</span></span>
<span>  <span class="va">optimal_k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectOptimalK.html">selectOptimalK</a></span><span class="op">(</span><span class="va">stability_results</span><span class="op">)</span></span>
<span>  <span class="va">final_programs</span> <span class="op">&lt;-</span> <span class="fu">extractConsensusPrograms</span><span class="op">(</span><span class="va">optimal_k</span><span class="op">)</span></span>
<span>  <span class="va">usage_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refitUsageMatrix.html">refitUsageMatrix</a></span><span class="op">(</span><span class="va">data</span>, <span class="va">final_programs</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="running-consensus-nmf">Running Consensus NMF<a class="anchor" aria-label="anchor" href="#running-consensus-nmf"></a>
</h2>
<div class="section level3">
<h3 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create example data with more structure for demonstration</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/mockSCE.html" class="external-link">mockSCE</a></span><span class="op">(</span>ngenes <span class="op">=</span> <span class="fl">1000</span>, ncells <span class="op">=</span> <span class="fl">800</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add cell type structure</span></span>
<span><span class="va">n_types</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Type"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">n_types</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Normalize</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add visualization coordinates</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run consensus NMF</span></span>
<span><span class="co"># Note: Using small parameters for vignette speed</span></span>
<span><span class="co"># In practice, use n_runs = 100-200 and broader k_range</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runConsensusNMF.html">runConsensusNMF</a></span><span class="op">(</span><span class="va">sce</span>, </span>
<span>                       k_range <span class="op">=</span> <span class="fl">4</span><span class="op">:</span><span class="fl">8</span>,           <span class="co"># Test k values</span></span>
<span>                       n_runs <span class="op">=</span> <span class="fl">20</span>,             <span class="co"># Number of runs per k</span></span>
<span>                       n_cores <span class="op">=</span> <span class="fl">1</span>,             <span class="co"># Parallel cores</span></span>
<span>                       verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check what was added</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "cNMF"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDimNames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "PCA"  "UMAP" "cNMF"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="parameter-selection">Parameter Selection<a class="anchor" aria-label="anchor" href="#parameter-selection"></a>
</h3>
<div class="section level4">
<h4 id="critical-parameters">Critical Parameters<a class="anchor" aria-label="anchor" href="#critical-parameters"></a>
</h4>
<p><strong>k_range</strong>: Range of k values to test - Start with
biologically expected range - Consider 2-3x expected number of cell
types - Broader ranges provide better stability assessment</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For 5 expected cell types, test:</span></span>
<span><span class="va">k_range</span> <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">12</span>    <span class="co"># Conservative range</span></span>
<span><span class="va">k_range</span> <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">15</span>    <span class="co"># Broader exploration</span></span></code></pre></div>
<p><strong>n_runs</strong>: Number of NMF runs per k value - More runs =
better stability estimates - Minimum: 50 runs for reasonable estimates -
Recommended: 100-200 runs for publication quality</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_runs</span> <span class="op">=</span> <span class="fl">50</span>      <span class="co"># Quick exploration</span></span>
<span><span class="va">n_runs</span> <span class="op">=</span> <span class="fl">100</span>     <span class="co"># Standard analysis  </span></span>
<span><span class="va">n_runs</span> <span class="op">=</span> <span class="fl">200</span>     <span class="co"># High confidence</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="algorithmic-parameters">Algorithmic Parameters<a class="anchor" aria-label="anchor" href="#algorithmic-parameters"></a>
</h4>
<p><strong>density_threshold</strong>: Outlier filtering stringency -
Higher values = more aggressive filtering - Range: 0.1 (lenient) to 0.8
(strict) - Default: 0.5 (balanced)</p>
<p><strong>n_cores</strong>: Parallel processing - Use available cores -
1 for system responsiveness - Scales nearly linearly with core count</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runConsensusNMF.html">runConsensusNMF</a></span><span class="op">(</span><span class="va">sce</span>,</span>
<span>                       k_range <span class="op">=</span> <span class="fl">5</span><span class="op">:</span><span class="fl">15</span>,</span>
<span>                       n_runs <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                       density_threshold <span class="op">=</span> <span class="fl">0.4</span>,    <span class="co"># Lenient filtering</span></span>
<span>                       n_cores <span class="op">=</span> <span class="fl">4</span>,                <span class="co"># Parallel processing</span></span>
<span>                       seed <span class="op">=</span> <span class="fl">123</span><span class="op">)</span>                 <span class="co"># Reproducibility</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="accessing-and-interpreting-results">Accessing and Interpreting Results<a class="anchor" aria-label="anchor" href="#accessing-and-interpreting-results"></a>
</h2>
<div class="section level3">
<h3 id="consensus-gene-expression-programs">Consensus Gene Expression Programs<a class="anchor" aria-label="anchor" href="#consensus-gene-expression-programs"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get consensus gene expression programs</span></span>
<span><span class="va">consensus_geps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getConsensusGEPs.html">getConsensusGEPs</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">consensus_geps</span><span class="op">)</span>  <span class="co"># genes × programs</span></span>
<span><span class="co">#&gt; [1] 1000    4</span></span>
<span></span>
<span><span class="co"># Get optimal k selected by algorithm</span></span>
<span><span class="va">optimal_k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getOptimalK.html">getOptimalK</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Optimal k selected:"</span>, <span class="va">optimal_k</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Optimal k selected: 4</span></span>
<span></span>
<span><span class="co"># Get top genes per program</span></span>
<span><span class="va">top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getTopGEPFeatures.html">getTopGEPFeatures</a></span><span class="op">(</span><span class="va">sce</span>, n <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>  <span class="co"># Top genes for program 1</span></span>
<span><span class="co">#&gt; [1] "Gene_0273" "Gene_0830" "Gene_0328" "Gene_0792" "Gene_0813" "Gene_0506"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="program-usage">Program Usage<a class="anchor" aria-label="anchor" href="#program-usage"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get program usage per cell</span></span>
<span><span class="va">gep_usage</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGEPUsage.html">getGEPUsage</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">gep_usage</span><span class="op">)</span>  <span class="co"># cells × programs</span></span>
<span><span class="co">#&gt; [1] 800   4</span></span>
<span></span>
<span><span class="co"># Usage is also stored in reducedDims</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">gep_usage</span>, <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"cNMF"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="stability-metrics">Stability Metrics<a class="anchor" aria-label="anchor" href="#stability-metrics"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get comprehensive stability metrics</span></span>
<span><span class="va">stability</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getStabilityMetrics.html">getStabilityMetrics</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">stability</span><span class="op">)</span></span>
<span><span class="co">#&gt;   k    stability silhouette reproducibility cophenetic_correlation</span></span>
<span><span class="co">#&gt; 1 4 0.0509493671  0.5022110      0.43750000                      1</span></span>
<span><span class="co">#&gt; 2 5 0.0307070707  0.6056696      0.40000000                      1</span></span>
<span><span class="co">#&gt; 3 6 0.0004201681  0.2913393      0.06666667                      1</span></span>
<span><span class="co">#&gt; 4 7 0.0009249743  0.4442262      0.10000000                      1</span></span>
<span><span class="co">#&gt; 5 8 0.0059748428  0.3627133      0.21250000                      1</span></span>
<span></span>
<span><span class="co"># Individual metrics explained:</span></span>
<span><span class="co"># - stability: clustering-based stability (higher = more stable)</span></span>
<span><span class="co"># - silhouette: silhouette score of consensus clustering</span></span>
<span><span class="co"># - reproducibility: how often same programs are recovered</span></span>
<span><span class="co"># - cophenetic_correlation: consensus matrix quality</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="accessing-different-k-values">Accessing Different k Values<a class="anchor" aria-label="anchor" href="#accessing-different-k-values"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Access results for specific k values</span></span>
<span><span class="va">available_k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAvailableK.html">getAvailableK</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Available k values:"</span>, <span class="va">available_k</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Available k values: 4 5 6 7 8</span></span>
<span></span>
<span><span class="co"># Get results for k=6 specifically</span></span>
<span><span class="va">geps_k6</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getConsensusGEPs.html">getConsensusGEPs</a></span><span class="op">(</span><span class="va">sce</span>, k <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="va">usage_k6</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGEPUsage.html">getGEPUsage</a></span><span class="op">(</span><span class="va">sce</span>, k <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">geps_k6</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1000    6</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">usage_k6</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 800   6</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="diagnostic-plots">Diagnostic Plots<a class="anchor" aria-label="anchor" href="#diagnostic-plots"></a>
</h2>
<div class="section level3">
<h3 id="stability-assessment">Stability Assessment<a class="anchor" aria-label="anchor" href="#stability-assessment"></a>
</h3>
<p>The most important diagnostic is the stability plot for k
selection:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot stability metrics across k values</span></span>
<span><span class="fu"><a href="../reference/plotStability.html">plotStability</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<p><img src="consensus-nmf_files/figure-html/plot-stability-1.png" width="960"></p>
<p><strong>Interpretation</strong>: - <strong>Stability</strong>:
Primary metric - choose k at peak or plateau -
<strong>Silhouette</strong>: Quality of consensus clustering -
<strong>Reproducibility</strong>: Consistency across runs -
<strong>Cophenetic</strong>: Consensus matrix block structure</p>
<p>Look for: - Peak in stability curve - High silhouette scores
(&gt;0.5) - Stable or increasing reproducibility - High cophenetic
correlation</p>
</div>
<div class="section level3">
<h3 id="gene-expression-program-heatmaps">Gene Expression Program Heatmaps<a class="anchor" aria-label="anchor" href="#gene-expression-program-heatmaps"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot gene expression programs</span></span>
<span><span class="fu"><a href="../reference/plotGEPs.html">plotGEPs</a></span><span class="op">(</span><span class="va">sce</span>, programs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="../reference/getOptimalK.html">getOptimalK</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, n_genes <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="consensus-nmf_files/figure-html/plot-geps-1.png" width="1152"></p>
<p><strong>Interpretation</strong>: - Each column = one gene expression
program - Each row = one gene (top contributors shown) - Red = high
expression, Blue = low expression - Look for distinct, interpretable
gene sets</p>
</div>
<div class="section level3">
<h3 id="program-usage-visualization">Program Usage Visualization<a class="anchor" aria-label="anchor" href="#program-usage-visualization"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot program usage on UMAP</span></span>
<span><span class="fu"><a href="../reference/plotGEPUsage.html">plotGEPUsage</a></span><span class="op">(</span><span class="va">sce</span>, programs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fu"><a href="../reference/getOptimalK.html">getOptimalK</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>, reduction <span class="op">=</span> <span class="st">"UMAP"</span><span class="op">)</span></span></code></pre></div>
<p><img src="consensus-nmf_files/figure-html/plot-usage-1.png" width="1152"></p>
<p><strong>Interpretation</strong>: - Each panel shows usage of one
program - Continuous colors = program activity levels - Spatial patterns
often reveal biological structures - Look for cell type or
state-specific patterns</p>
</div>
</div>
<div class="section level2">
<h2 id="advanced-analysis">Advanced Analysis<a class="anchor" aria-label="anchor" href="#advanced-analysis"></a>
</h2>
<div class="section level3">
<h3 id="manual-k-selection">Manual k Selection<a class="anchor" aria-label="anchor" href="#manual-k-selection"></a>
</h3>
<p>Sometimes you may want to override automatic k selection:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get detailed information about all k values</span></span>
<span><span class="va">cnmf_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getConsensusNMFInfo.html">getConsensusNMFInfo</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">cnmf_info</span><span class="op">)</span></span>
<span><span class="co">#&gt; $k_range</span></span>
<span><span class="co">#&gt; [1] 4 5 6 7 8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $n_runs</span></span>
<span><span class="co">#&gt; [1] 20</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $optimal_k</span></span>
<span><span class="co">#&gt; [1] 4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $assay_used</span></span>
<span><span class="co">#&gt; [1] "logcounts"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $subset_row</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $parameters</span></span>
<span><span class="co">#&gt; $parameters$tol</span></span>
<span><span class="co">#&gt; [1] 1e-05</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $parameters$maxit</span></span>
<span><span class="co">#&gt; [1] 100</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $parameters$L1</span></span>
<span><span class="co">#&gt; [1] 0 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $parameters$n_cores</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $available_k_values</span></span>
<span><span class="co">#&gt; [1] "4" "5" "6" "7" "8"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $stability_summary</span></span>
<span><span class="co">#&gt;        k       stability           silhouette     reproducibility  </span></span>
<span><span class="co">#&gt;  Min.   :4   Min.   :0.0004202   Min.   :0.2913   Min.   :0.06667  </span></span>
<span><span class="co">#&gt;  1st Qu.:5   1st Qu.:0.0009250   1st Qu.:0.3627   1st Qu.:0.10000  </span></span>
<span><span class="co">#&gt;  Median :6   Median :0.0059748   Median :0.4442   Median :0.21250  </span></span>
<span><span class="co">#&gt;  Mean   :6   Mean   :0.0177953   Mean   :0.4412   Mean   :0.24333  </span></span>
<span><span class="co">#&gt;  3rd Qu.:7   3rd Qu.:0.0307071   3rd Qu.:0.5022   3rd Qu.:0.40000  </span></span>
<span><span class="co">#&gt;  Max.   :8   Max.   :0.0509494   Max.   :0.6057   Max.   :0.43750  </span></span>
<span><span class="co">#&gt;  cophenetic_correlation</span></span>
<span><span class="co">#&gt;  Min.   :1             </span></span>
<span><span class="co">#&gt;  1st Qu.:1             </span></span>
<span><span class="co">#&gt;  Median :1             </span></span>
<span><span class="co">#&gt;  Mean   :1             </span></span>
<span><span class="co">#&gt;  3rd Qu.:1             </span></span>
<span><span class="co">#&gt;  Max.   :1             </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $call</span></span>
<span><span class="co">#&gt; runConsensusNMF(x = sce, k_range = 4:8, n_runs = 20, verbose = TRUE, </span></span>
<span><span class="co">#&gt;     n_cores = 1)</span></span>
<span></span>
<span><span class="co"># Force usage of specific k (e.g., k=6)</span></span>
<span><span class="co"># This changes what getConsensusGEPs() returns by default</span></span>
<span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">cNMF</span><span class="op">$</span><span class="va">cNMF</span><span class="op">$</span><span class="va">optimal_k</span> <span class="op">&lt;-</span> <span class="fl">6</span></span>
<span></span>
<span><span class="co"># Verify change</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"New optimal k:"</span>, <span class="fu"><a href="../reference/getOptimalK.html">getOptimalK</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="co">#&gt; New optimal k: 6</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="program-annotation">Program Annotation<a class="anchor" aria-label="anchor" href="#program-annotation"></a>
</h3>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get top genes for biological annotation</span></span>
<span><span class="va">top_genes_detailed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getTopGEPFeatures.html">getTopGEPFeatures</a></span><span class="op">(</span><span class="va">sce</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Example: annotate programs based on top genes</span></span>
<span><span class="va">opt_k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getOptimalK.html">getOptimalK</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">program_annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="va">opt_k</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">opt_k</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">top_10</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">top_genes_detailed</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="fl">10</span><span class="op">)</span></span>
<span>  <span class="va">program_annotations</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Program_"</span>, <span class="va">i</span>, <span class="st">" ("</span>, </span>
<span>                                  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">top_10</span>, <span class="fl">3</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span>, </span>
<span>                                  <span class="st">"...)"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">program_annotations</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Program_1 (Gene_0830, Gene_0273, Gene_0328...)"</span></span>
<span><span class="co">#&gt; [2] "Program_2 (Gene_0830, Gene_0341, Gene_0328...)"</span></span>
<span><span class="co">#&gt; [3] "Program_3 (Gene_0830, Gene_0341, Gene_0273...)"</span></span>
<span><span class="co">#&gt; [4] "Program_4 (Gene_0112, Gene_0273, Gene_0527...)"</span></span>
<span><span class="co">#&gt; [5] "Program_5 (Gene_0830, Gene_0341, Gene_0136...)"</span></span>
<span><span class="co">#&gt; [6] "Program_6 (Gene_0830, Gene_0341, Gene_0273...)"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="robustness-assessment">Robustness Assessment<a class="anchor" aria-label="anchor" href="#robustness-assessment"></a>
</h3>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test robustness with different parameters</span></span>
<span><span class="va">sce_alt1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runConsensusNMF.html">runConsensusNMF</a></span><span class="op">(</span><span class="va">sce</span>, k_range <span class="op">=</span> <span class="fl">4</span><span class="op">:</span><span class="fl">8</span>, n_runs <span class="op">=</span> <span class="fl">50</span>, </span>
<span>                           density_threshold <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span><span class="va">sce_alt2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runConsensusNMF.html">runConsensusNMF</a></span><span class="op">(</span><span class="va">sce</span>, k_range <span class="op">=</span> <span class="fl">4</span><span class="op">:</span><span class="fl">8</span>, n_runs <span class="op">=</span> <span class="fl">50</span>, </span>
<span>                           density_threshold <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare optimal k selections</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>original <span class="op">=</span> <span class="fu"><a href="../reference/getOptimalK.html">getOptimalK</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,</span>
<span>  lenient <span class="op">=</span> <span class="fu"><a href="../reference/getOptimalK.html">getOptimalK</a></span><span class="op">(</span><span class="va">sce_alt1</span><span class="op">)</span>,</span>
<span>  strict <span class="op">=</span> <span class="fu"><a href="../reference/getOptimalK.html">getOptimalK</a></span><span class="op">(</span><span class="va">sce_alt2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="computational-considerations">Computational Considerations<a class="anchor" aria-label="anchor" href="#computational-considerations"></a>
</h2>
<div class="section level3">
<h3 id="performance-scaling">Performance Scaling<a class="anchor" aria-label="anchor" href="#performance-scaling"></a>
</h3>
<p>cNMF is computationally intensive. Runtime scales with:</p>
<ul>
<li>
<strong>n_runs</strong>: Linear scaling (100 runs = 2x time vs 50
runs)</li>
<li>
<strong>k_range</strong>: Linear scaling (each k adds ~n_runs
time)</li>
<li>
<strong>Data size</strong>: Quadratic in min(genes, cells)</li>
<li>
<strong>n_cores</strong>: Near-linear speedup with
parallelization</li>
</ul>
</div>
<div class="section level3">
<h3 id="memory-requirements">Memory Requirements<a class="anchor" aria-label="anchor" href="#memory-requirements"></a>
</h3>
<p>Memory usage depends on: - Storing all intermediate NMF results -
Consensus matrix construction - Peak usage during clustering steps</p>
<p>For large datasets (&gt;50k cells), consider: - Reducing n_runs
(minimum 50) - Narrower k_range based on prior knowledge - Using
high-memory computing environments</p>
</div>
<div class="section level3">
<h3 id="optimization-tips">Optimization Tips<a class="anchor" aria-label="anchor" href="#optimization-tips"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For large datasets</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runConsensusNMF.html">runConsensusNMF</a></span><span class="op">(</span><span class="va">sce</span>,</span>
<span>                       k_range <span class="op">=</span> <span class="fl">8</span><span class="op">:</span><span class="fl">12</span>,          <span class="co"># Focused range</span></span>
<span>                       n_runs <span class="op">=</span> <span class="fl">75</span>,             <span class="co"># Reduced runs</span></span>
<span>                       n_cores <span class="op">=</span> <span class="fl">8</span>,             <span class="co"># Max parallelization</span></span>
<span>                       density_threshold <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span> <span class="co"># Aggressive filtering</span></span>
<span></span>
<span><span class="co"># For exploration with quick turnaround</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runConsensusNMF.html">runConsensusNMF</a></span><span class="op">(</span><span class="va">sce</span>,</span>
<span>                       k_range <span class="op">=</span> <span class="fl">5</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>                       n_runs <span class="op">=</span> <span class="fl">25</span>,             <span class="co"># Very fast</span></span>
<span>                       n_cores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="integration-with-downstream-analysis">Integration with Downstream Analysis<a class="anchor" aria-label="anchor" href="#integration-with-downstream-analysis"></a>
</h2>
<div class="section level3">
<h3 id="cell-type-assignment">Cell Type Assignment<a class="anchor" aria-label="anchor" href="#cell-type-assignment"></a>
</h3>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assign cells to dominant programs</span></span>
<span><span class="va">usage_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGEPUsage.html">getGEPUsage</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">dominant_programs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">usage_matrix</span>, <span class="fl">1</span>, <span class="va">which.max</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add to cell metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">dominant_program</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Program_"</span>, <span class="va">dominant_programs</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare with original cell types (if available)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">dominant_program</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">cell_type</span><span class="op">)</span></span>
<span><span class="co">#&gt;            </span></span>
<span><span class="co">#&gt;             Type1 Type2 Type3 Type4 Type5</span></span>
<span><span class="co">#&gt;   Program_1    38    34    38    38    37</span></span>
<span><span class="co">#&gt;   Program_2    49    47    52    51    59</span></span>
<span><span class="co">#&gt;   Program_3    42    47    56    49    52</span></span>
<span><span class="co">#&gt;   Program_4    27    24    19    23    18</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="differential-program-usage">Differential Program Usage<a class="anchor" aria-label="anchor" href="#differential-program-usage"></a>
</h3>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test for differential program usage between conditions</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Example: differential usage between cell types</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">program_1_usage</span> <span class="op">&lt;-</span> <span class="va">usage_matrix</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Statistical testing</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers</a></span><span class="op">(</span><span class="va">sce</span>, </span>
<span>                      groups <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">cell_type</span>,</span>
<span>                      test.type <span class="op">=</span> <span class="st">"t"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="gene-set-enrichment">Gene Set Enrichment<a class="anchor" aria-label="anchor" href="#gene-set-enrichment"></a>
</h3>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example pathway enrichment (requires additional packages)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get top genes for program 1</span></span>
<span><span class="va">program_1_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getTopGEPFeatures.html">getTopGEPFeatures</a></span><span class="op">(</span><span class="va">sce</span>, n <span class="op">=</span> <span class="fl">100</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Run enrichment analysis</span></span>
<span><span class="va">enrichment</span> <span class="op">&lt;-</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">program_1_genes</span>,</span>
<span>                      OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>                      ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>                      readable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="best-practices">Best Practices<a class="anchor" aria-label="anchor" href="#best-practices"></a>
</h2>
<div class="section level3">
<h3 id="parameter-selection-guidelines">Parameter Selection Guidelines<a class="anchor" aria-label="anchor" href="#parameter-selection-guidelines"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<strong>Start broad</strong>: Use wide k_range for initial
exploration</li>
<li>
<strong>Sufficient runs</strong>: Minimum 50 runs, prefer 100+ for
final analysis</li>
<li>
<strong>Multiple thresholds</strong>: Test different
density_threshold values</li>
<li>
<strong>Reproducibility</strong>: Use set.seed() for reproducible
results</li>
</ol>
</div>
<div class="section level3">
<h3 id="quality-control">Quality Control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<strong>Stability curves</strong>: Should show clear peaks or
plateaus</li>
<li>
<strong>Program interpretation</strong>: Programs should be
biologically meaningful</li>
<li>
<strong>Usage patterns</strong>: Should correlate with known cell
types/states</li>
<li>
<strong>Consistency</strong>: Results should be robust to parameter
changes</li>
</ol>
</div>
<div class="section level3">
<h3 id="common-pitfalls">Common Pitfalls<a class="anchor" aria-label="anchor" href="#common-pitfalls"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<strong>Too few runs</strong>: &lt;50 runs can give unstable
results</li>
<li>
<strong>Narrow k_range</strong>: May miss optimal k value</li>
<li>
<strong>Over-interpretation</strong>: Programs may not always
correspond to cell types</li>
<li>
<strong>Ignoring stability</strong>: Don’t just pick k based on
expectations</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="comparison-with-standard-nmf">Comparison with Standard NMF<a class="anchor" aria-label="anchor" href="#comparison-with-standard-nmf"></a>
</h2>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run standard NMF with same k</span></span>
<span><span class="va">opt_k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getOptimalK.html">getOptimalK</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runNMF.html">runNMF</a></span><span class="op">(</span><span class="va">sce</span>, k <span class="op">=</span> <span class="va">opt_k</span>, name <span class="op">=</span> <span class="st">"StandardNMF"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare stability by running multiple times</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">nmf_runs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">replicate</a></span><span class="op">(</span><span class="fl">10</span>, <span class="op">{</span></span>
<span>  <span class="va">sce_temp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runNMF.html">runNMF</a></span><span class="op">(</span><span class="va">sce</span>, k <span class="op">=</span> <span class="va">opt_k</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/getBasis.html">getBasis</a></span><span class="op">(</span><span class="va">sce_temp</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>  <span class="co"># First program</span></span>
<span><span class="op">}</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate coefficient of variation across runs</span></span>
<span><span class="va">cv_standard</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="va">nmf_runs</span><span class="op">)</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">sd</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare with cNMF stability (programs should be more stable)</span></span>
<span><span class="va">cnmf_program_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getConsensusGEPs.html">getConsensusGEPs</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Standard NMF variability (mean CV):"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">cv_standard</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"cNMF provides stable programs across runs\n"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.3 Patched (2025-02-28 r87922)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Sequoia 15.6.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/New_York</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] scran_1.34.0                pheatmap_1.0.12            </span></span>
<span><span class="co">#&gt;  [3] scater_1.34.1               ggplot2_3.5.1              </span></span>
<span><span class="co">#&gt;  [5] scuttle_1.16.0              BiocNMF_0.99.0             </span></span>
<span><span class="co">#&gt;  [7] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0</span></span>
<span><span class="co">#&gt;  [9] Biobase_2.66.0              GenomicRanges_1.58.0       </span></span>
<span><span class="co">#&gt; [11] GenomeInfoDb_1.42.3         IRanges_2.40.1             </span></span>
<span><span class="co">#&gt; [13] S4Vectors_0.44.0            MatrixGenerics_1.18.1      </span></span>
<span><span class="co">#&gt; [15] matrixStats_1.5.0           BiocGenerics_0.52.0        </span></span>
<span><span class="co">#&gt; [17] BiocStyle_2.34.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] gridExtra_2.3           rlang_1.1.5             magrittr_2.0.3         </span></span>
<span><span class="co">#&gt;  [4] compiler_4.4.3          systemfonts_1.2.1       vctrs_0.6.5            </span></span>
<span><span class="co">#&gt;  [7] pkgconfig_2.0.3         crayon_1.5.3            fastmap_1.2.0          </span></span>
<span><span class="co">#&gt; [10] XVector_0.46.0          labeling_0.4.3          rmarkdown_2.29         </span></span>
<span><span class="co">#&gt; [13] UCSC.utils_1.2.0        ggbeeswarm_0.7.2        ragg_1.5.0             </span></span>
<span><span class="co">#&gt; [16] xfun_0.51               bluster_1.16.0          zlibbioc_1.52.0        </span></span>
<span><span class="co">#&gt; [19] cachem_1.1.0            beachmat_2.22.0         jsonlite_1.9.1         </span></span>
<span><span class="co">#&gt; [22] DelayedArray_0.32.0     BiocParallel_1.40.0     irlba_2.3.5.1          </span></span>
<span><span class="co">#&gt; [25] parallel_4.4.3          cluster_2.1.8.1         R6_2.6.1               </span></span>
<span><span class="co">#&gt; [28] bslib_0.9.0             RColorBrewer_1.1-3      limma_3.62.2           </span></span>
<span><span class="co">#&gt; [31] jquerylib_0.1.4         Rcpp_1.0.14             bookdown_0.42          </span></span>
<span><span class="co">#&gt; [34] knitr_1.49              FNN_1.1.4.1             Matrix_1.7-3           </span></span>
<span><span class="co">#&gt; [37] igraph_2.1.4            tidyselect_1.2.1        abind_1.4-8            </span></span>
<span><span class="co">#&gt; [40] yaml_2.3.10             viridis_0.6.5           codetools_0.2-20       </span></span>
<span><span class="co">#&gt; [43] lattice_0.22-6          tibble_3.2.1            withr_3.0.2            </span></span>
<span><span class="co">#&gt; [46] evaluate_1.0.3          desc_1.4.3              pillar_1.10.1          </span></span>
<span><span class="co">#&gt; [49] BiocManager_1.30.25     generics_0.1.3          munsell_0.5.1          </span></span>
<span><span class="co">#&gt; [52] scales_1.3.0            glue_1.8.0              metapod_1.14.0         </span></span>
<span><span class="co">#&gt; [55] tools_4.4.3             BiocNeighbors_2.0.1     ScaledMatrix_1.14.0    </span></span>
<span><span class="co">#&gt; [58] locfit_1.5-9.12         RcppML_0.3.7            fs_1.6.5               </span></span>
<span><span class="co">#&gt; [61] grid_4.4.3              edgeR_4.4.2             colorspace_2.1-1       </span></span>
<span><span class="co">#&gt; [64] GenomeInfoDbData_1.2.13 beeswarm_0.4.0          BiocSingular_1.22.0    </span></span>
<span><span class="co">#&gt; [67] vipor_0.4.7             cli_3.6.4               rsvd_1.0.5             </span></span>
<span><span class="co">#&gt; [70] textshaping_1.0.0       S4Arrays_1.6.0          viridisLite_0.4.2      </span></span>
<span><span class="co">#&gt; [73] dplyr_1.1.4             uwot_0.2.3              gtable_0.3.6           </span></span>
<span><span class="co">#&gt; [76] sass_0.4.9              digest_0.6.37           SparseArray_1.6.2      </span></span>
<span><span class="co">#&gt; [79] ggrepel_0.9.6           dqrng_0.4.1             htmlwidgets_1.6.4      </span></span>
<span><span class="co">#&gt; [82] farver_2.1.2            htmltools_0.5.8.1       pkgdown_2.1.3          </span></span>
<span><span class="co">#&gt; [85] lifecycle_1.0.4         httr_1.4.7              statmod_1.5.0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ol style="list-style-type: decimal">
<li><p>Kotliar D, Veres A, Nagy MA, et al. Identifying gene expression
programs of cell-type identity and cellular activity with single-cell
RNA-Seq. <em>eLife</em>. 2019;8:e43803.</p></li>
<li><p>DeBruine ZJ, Melcher K, Triche TJ. High-performance non-negative
matrix factorization for large single cell data. <em>bioRxiv</em>.
2021.</p></li>
<li><p>Lee DD, Seung HS. Learning the parts of objects by non-negative
matrix factorization. <em>Nature</em>. 1999;401(6755):788-791.</p></li>
</ol>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/MicTott" class="external-link">Michael Totty</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
