<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ BiocNMF</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BiocNMF</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/BiocNMF.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/getting-started.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/consensus-nmf.html">Consensus NMF</a></li>
    <li><a class="dropdown-item" href="../articles/go-enrichment.html">GO Enrichment Analysis</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MicTott/BiocNMF" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="main-functions">Main Functions<a class="anchor" aria-label="anchor" href="#main-functions"></a></h2>

      <div class="section-desc"><p>Core NMF and consensus NMF functions</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="runBiocNMF.html">runBiocNMF()</a></code>

        </dt>
        <dd>Run BiocNMF Non-negative Matrix Factorization on SingleCellExperiment objects</dd>
      </dl><dl><dt>

          <code><a href="runConsensusNMF.html">runConsensusNMF()</a></code>

        </dt>
        <dd>Run Consensus Non-negative Matrix Factorization on SingleCellExperiment objects</dd>
      </dl></div><div class="section level2">
      <h2 id="standard-nmf-accessors">Standard NMF Accessors<a class="anchor" aria-label="anchor" href="#standard-nmf-accessors"></a></h2>

      <div class="section-desc"><p>Access standard NMF results</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getBasis.html">getBasis()</a></code>

        </dt>
        <dd>Extract NMF basis matrix from SingleCellExperiment object</dd>
      </dl><dl><dt>

          <code><a href="getCoefficients.html">getCoefficients()</a></code>

        </dt>
        <dd>Extract NMF coefficient matrix from SingleCellExperiment object</dd>
      </dl><dl><dt>

          <code><a href="getTopFeatures.html">getTopFeatures()</a></code>

        </dt>
        <dd>Calculate feature loadings (which genes contribute most to each factor)</dd>
      </dl><dl><dt>

          <code><a href="reconstructNMF.html">reconstructNMF()</a></code>

        </dt>
        <dd>Reconstruct original matrix from NMF factors</dd>
      </dl></div><div class="section level2">
      <h2 id="consensus-nmf-accessors">Consensus NMF Accessors<a class="anchor" aria-label="anchor" href="#consensus-nmf-accessors"></a></h2>

      <div class="section-desc"><p>Access consensus NMF results</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getConsensusGEPs.html">getConsensusGEPs()</a></code>

        </dt>
        <dd>Extract consensus gene expression programs (GEPs) from cNMF results</dd>
      </dl><dl><dt>

          <code><a href="getGEPUsage.html">getGEPUsage()</a></code>

        </dt>
        <dd>Extract GEP usage matrix from cNMF results</dd>
      </dl><dl><dt>

          <code><a href="getStabilityMetrics.html">getStabilityMetrics()</a></code>

        </dt>
        <dd>Get stability metrics from cNMF results</dd>
      </dl><dl><dt>

          <code><a href="getOptimalK.html">getOptimalK()</a></code>

        </dt>
        <dd>Get optimal k value from cNMF results</dd>
      </dl><dl><dt>

          <code><a href="getTopGEPFeatures.html">getTopGEPFeatures()</a></code>

        </dt>
        <dd>Get top features for each consensus gene expression program</dd>
      </dl><dl><dt>

          <code><a href="getConsensusNMFInfo.html">getConsensusNMFInfo()</a></code>

        </dt>
        <dd>Get consensus NMF information and parameters</dd>
      </dl><dl><dt>

          <code><a href="getAvailableK.html">getAvailableK()</a></code>

        </dt>
        <dd>Get available k values from cNMF results</dd>
      </dl></div><div class="section level2">
      <h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a></h2>

      <div class="section-desc"><p>Plotting functions for diagnostics and results</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="vizUMAP.html">vizUMAP()</a></code>

        </dt>
        <dd>Visualize NMF programs on UMAP coordinates</dd>
      </dl><dl><dt>

          <code><a href="vizDimRed.html">vizDimRed()</a></code>

        </dt>
        <dd>Visualize NMF programs on dimension reduction plots</dd>
      </dl><dl><dt>

          <code><a href="plotStability.html">plotStability()</a></code>

        </dt>
        <dd>Plot stability metrics from consensus NMF</dd>
      </dl><dl><dt>

          <code><a href="plotStabilityComparison.html">plotStabilityComparison()</a></code>

        </dt>
        <dd>Plot comparison of stability metrics across different k values</dd>
      </dl><dl><dt>

          <code><a href="plotGEPs.html">plotGEPs()</a></code>

        </dt>
        <dd>Plot gene expression programs (GEPs) heatmap</dd>
      </dl><dl><dt>

          <code><a href="plotGEPUsage.html">plotGEPUsage()</a></code>

        </dt>
        <dd>Plot GEP usage in reduced dimension space</dd>
      </dl></div><div class="section level2">
      <h2 id="gene-analysis-enrichment">Gene Analysis &amp; Enrichment<a class="anchor" aria-label="anchor" href="#gene-analysis-enrichment"></a></h2>

      <div class="section-desc"><p>Functions for gene analysis and pathway enrichment</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="FindAllDEPs.html">FindAllDEPs()</a></code>

        </dt>
        <dd>Find All Differentially Expressed Programs (1-vs-all analysis)</dd>
      </dl><dl><dt>

          <code><a href="plotDEPsVolcano.html">plotDEPsVolcano()</a></code>

        </dt>
        <dd>Visualize Differentially Expressed Programs with volcano plots</dd>
      </dl><dl><dt>

          <code><a href="plotDEPsDots.html">plotDEPsDots()</a></code>

        </dt>
        <dd>Visualize Differentially Expressed Programs with dot plot</dd>
      </dl><dl><dt>

          <code><a href="plotProgramGenes.html">plotProgramGenes()</a></code>

        </dt>
        <dd>Visualize top genes per NMF program with heatmap</dd>
      </dl><dl><dt>

          <code><a href="prepareProgramGeneSets.html">prepareProgramGeneSets()</a></code>

        </dt>
        <dd>Prepare gene sets for enrichment analysis from NMF programs</dd>
      </dl><dl><dt>

          <code><a href="runProgramGOEnrichment.html">runProgramGOEnrichment()</a></code>

        </dt>
        <dd>Run Gene Ontology enrichment on NMF program gene sets</dd>
      </dl><dl><dt>

          <code><a href="plotGOBarChart.html">plotGOBarChart()</a></code>

        </dt>
        <dd>Plot GO Enrichment Bar Chart</dd>
      </dl><dl><dt>

          <code><a href="plotGODotPlot.html">plotGODotPlot()</a></code>

        </dt>
        <dd>Plot GO Enrichment Dot Plot</dd>
      </dl><dl><dt>

          <code><a href="plotMultipleProgramsGO.html">plotMultipleProgramsGO()</a></code>

        </dt>
        <dd>Plot Multiple Programs GO Comparison</dd>
      </dl><dl><dt>

          <code><a href="plotGONetwork.html">plotGONetwork()</a></code>

        </dt>
        <dd>Plot GO Network</dd>
      </dl><dl><dt>

          <code><a href="saveGOPlotsToPDF.html">saveGOPlotsToPDF()</a></code>

        </dt>
        <dd>Save GO Plots to PDF</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/MicTott" class="external-link">Michael Totty</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

